### PR TITLE
Create Feature Script

### PR DESCRIPTION
**Pull request by alexgraffeocohen**
This branch adds a create feature script that is very similar to the create hotfix one. It adds a feature story with 0 points based on the name you pass it. I also did a series of refactors so that I could easily drop in the new script and corresponding logic.

This script was built under similar assumptions as the hotfix one: minimum configuration/arguments necessary, sensible base-line defaults. So this does not let you pass custom points for stories or even tags, just like the hotfix one doesn't let you add tags either. In the future, if we want both of these scripts to let us add more options, I would want us to move to Thor to better handle that. http://whatisthor.com/

### CODE REVIEW

**Comment by danbernier** [comment]
What is Thor? Baby, don't hurt me, don't hurt me, no more...

On Tue, Nov 10, 2015 at 4:16 PM, Alex Wilkinson notifications@github.com
wrote:

> This branch adds a create feature script that is very similar to the
> create hotfix one. It adds a feature story with 0 points based on the name
> you pass it. I also did a series of refactors so that I could easily drop
> in the new script and corresponding logic.
> 
> This script was built under similar assumptions as the hotfix one: minimum
> configuration/arguments necessary, sensible base-line defaults. So this
> does not let you pass custom points for stories or even tags, just like the
> hotfix one doesn't let you add tags either. In the future, if we want both
> of these scripts to let us add more options, I would want us to move to
> 
> ## Thor to better handle that. http://whatisthor.com/
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/ContinuityControl/dotfiles/pull/36
> Commit Summary
> - Refactor: extract generic story creation logic to Story class
> - Refactor: inline-method (create_hotfix_name)
> - Refactor: rename-method (create_hotfix_name -> branch_name)
> - Refactor: move Hotfix#branch_name to Story
> - Add Create Feature script
> 
> File Changes
> - _A_ home/.bin/create-feature
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-0
>   (54)
> - _A_ home/.bin/feature.rb
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-1
>   (26)
> - _M_ home/.bin/git_repository.rb
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-2
>   (12)
> - _M_ home/.bin/hotfix.rb
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-3
>   (54)
> - _M_ home/.bin/pivotal_project.rb
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-4
>   (9)
> - _A_ home/.bin/story.rb
>   https://github.com/ContinuityControl/dotfiles/pull/36/files#diff-5
>   (52)
> 
> Patch Links:
> - https://github.com/ContinuityControl/dotfiles/pull/36.patch
> - https://github.com/ContinuityControl/dotfiles/pull/36.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ContinuityControl/dotfiles/pull/36.

## 

## 

---

This email and all files transmitted with it are intended solely for use by 
the intended recipient.  Any other use is prohibited.  If you received this 
email in error, please notify the sender, delete the email and destroy all 
attachments.  Opinions and ideas expressed herein are solely those of the 
sender, and may not reflect the official position of Continuity.

File: home/.bin/create-feature [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env ruby
+require_relative 'feature'
+
+help_text = <<-eoht
```
Comment (treznick) [inline home/.bin/create-feature:4]: convention here is that if there are any flags you can enumerate them here, but this is grood enough

**Comment by alexgraffeocohen** [comment]
:smile:

**Comment by danbernier** [comment]
Wow, there's lots more code in here than I realized! We should totally extract this to a gem, w/ Thor, and open source it. It might be easier to manage. (Don't take me at my word and go do it - that's extra work, there.)

File: home/.bin/create-feature (hunk 1 above)
Comment (alexgraffeocohen) [inline home/.bin/create-feature:4]: right. for now we just have the one flag, and I don't want to add more unless we want to get real serious

File: home/.bin/git_repository.rb [hunk 2]
```diff
@@ -25,6 +25,18 @@ def validate_for_hotfix
     end
   end
 
+  def validate_for_feature
```
Comment (treznick) [inline home/.bin/git_repository.rb]: maybe rename to `validate_for_feature!`?

**Comment by alexgraffeocohen** [comment]
@danbernier Yeah I guess this did all amount to quite a bit! I would love the excuse to learn to make a gem with this stuff. You or @treznick wanna help me out?

File: home/.bin/story.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @story_name = story_name
+    @project_id = project_id
+    @pivotal_token = pivotal_token
+  end
+
+  private
+
+  attr_reader :story_name, :project_id, :pivotal_token
+
+  def validate_environment_and_arguments
```
Comment (treznick) [inline home/.bin/story.rb]: again maybe add a `!`?

File: home/.bin/git_repository.rb (hunk 2 above)
Comment (alexgraffeocohen) [inline home/.bin/git_repository.rb]: I like that idea

**Comment by danbernier** [comment]
Alex: I can chat w/ you, sure!

On Tue, Nov 10, 2015 at 4:27 PM, Alex Wilkinson notifications@github.com
wrote:

> @danbernier https://github.com/danbernier Yeah I guess this did all
> amount to quite a bit! I would love the excuse to learn to make a gem with
> this stuff. You or @treznick https://github.com/treznick wanna help me
> out?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ContinuityControl/dotfiles/pull/36#issuecomment-155572421
> .

## 

## 

---

This email and all files transmitted with it are intended solely for use by 
the intended recipient.  Any other use is prohibited.  If you received this 
email in error, please notify the sender, delete the email and destroy all 
attachments.  Opinions and ideas expressed herein are solely those of the 
sender, and may not reflect the official position of Continuity.

File: home/.bin/story.rb (hunk 3 above)
Comment (alexgraffeocohen) [inline home/.bin/story.rb]: yeah starting to think more of these methods can have `!`, because they cause exceptions if things are broke

**Comment by alexgraffeocohen** [comment]
@danbernier awesome! part of the reason I wanted to start getting code out of the script files themselves was so that it could be more manageable in the present and easily transferrable to a Thor setup. So if we think there's value, Gem sounds like fun. I would love to be able to pass extra arguments, anyway

**Comment by alexgraffeocohen** [comment]
Mergeable for now pending small method renames?

**Comment by danbernier** [comment]
Fine by me, I haven't really looked at any of it. #yolo :shipit:

On Tue, Nov 10, 2015 at 4:36 PM, Alex Wilkinson notifications@github.com
wrote:

> Mergeable for now pending small method renamed?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ContinuityControl/dotfiles/pull/36#issuecomment-155574493
> .

## 

## 

---

This email and all files transmitted with it are intended solely for use by 
the intended recipient.  Any other use is prohibited.  If you received this 
email in error, please notify the sender, delete the email and destroy all 
attachments.  Opinions and ideas expressed herein are solely those of the 
sender, and may not reflect the official position of Continuity.

**Comment by alexgraffeocohen** [comment]
Yeah Dotfiles is kind of #yolo in general

File: home/.bin/story.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  def env_variable_undefined?(variable)
+    variable.nil? || variable.empty?
+  end
+
+  def pivotal_project
+    PivotalProject.new(pivotal_token, project_id)
+  end
+
+  def derive_story_owners
```
Comment (jhirbour) [inline home/.bin/story.rb]: If/When this becomes a gem: if there's not hitch file maybe scrape the users name out of the env? or an environment variable for "your pivotal username"...

**Pull request closed without merging**